executor/plan_recreator.go

```diff
@@ -0,0 +1,456 @@
+				err := os.Remove(recreatorPath + "/" + k)
+				if err != nil {
+					logutil.BgLogger().Warn(fmt.Sprintf("Cleaning outdated file %s failed.", k))
+				}
+				delete(Flist, k)
+				delete(TList, v.Token)
+			}
+		}
+	}
+	// Generate Token
+	token := md5.Sum([]byte(fmt.Sprintf("%s%d", fileName, rand.Int63())))
+	e.Ctx.Value(PlanRecreatorFileList).(fileList).FileInfo[fileName] = fileInfo{StartTime: startTime, Token: token}
+	e.Ctx.Value(PlanRecreatorFileList).(fileList).TokenMap[token] = fileName
+
+	// Create zip writer
+	zw := zip.NewWriter(zf)
+	defer func() {
+		err := zw.Close()
+		if err != nil {
+			logutil.BgLogger().Warn("Closing zip writer failed.")
+		}
+		err = zf.Close()
+		if err != nil {
+			logutil.BgLogger().Warn("Closing zip file failed.")
+		}
+	}()
+
+	// Dump config
+	cf, err := zw.Create("config.toml")
+	if err != nil {
+		return "", err
+	}
+	if err := toml.NewEncoder(cf).Encode(config.GetGlobalConfig()); err != nil {
+		return "", err
+	}
+
+	// Dump meta
+	mt, err := zw.Create("meta.txt")
+	if err != nil {
+		return "", err
+	}
+	_, err = mt.Write([]byte(printer.GetTiDBInfo()))
+	if err != nil {
+		return "", err
+	}
+
+	// Retrieve current DB
+	sessionVars := e.Ctx.GetSessionVars()
+	dbName := model.NewCIStr(sessionVars.CurrentDB)
+	do := domain.GetDomain(e.Ctx)
+
+	// Retrieve all tables
+	pairs, err := extractTableNames(e.ExecStmt, dbName.L)
+	if err != nil {
+		return "", errors.New(fmt.Sprintf("Plan Recreator: invalid SQL text, err: %v", err))
+	}
+
+	// Dump stats
+	for pair := range pairs {
+		jsonTbl, err := getStatsForTable(do, pair)
+		if err != nil {
+			return "", err
+		}
+		statsFw, err := zw.Create(fmt.Sprintf("stats/%v.%v.json", pair.DBName, pair.TableName))
+		if err != nil {
+			return "", err
+		}
+		data, err := json.Marshal(jsonTbl)
+		if err != nil {
+			return "", err
+		}
+		_, err = statsFw.Write(data)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// Dump schema
+	for pair := range pairs {
+		err = getShowCreateTable(pair, zw, e.Ctx)
+		if err != nil {
+			return "", err
+		}
+	}
+
+	// Dump variables
+	varMap := make(map[string]string)
+	recordSets, err := e.Ctx.(sqlexec.SQLExecutor).Execute(context.TODO(), "show variables")
+	if err != nil {
+		return "", err
+	}
+	sRows, err := resultSetToStringSlice(context.Background(), recordSets[0])
+	if err != nil {
+		return "", err
+	}
+	vf, err := zw.Create("variables.toml")
+	if err != nil {
+		return "", err
+	}
+	for _, row := range sRows {
+		varMap[row[0]] = row[1]
+	}
+	if err := toml.NewEncoder(vf).Encode(varMap); err != nil {
+		return "", err
+	}
+	if len(recordSets) > 0 {
+		if err := recordSets[0].Close(); err != nil {
+			return "", err
+		}
+	}
+
+	// Dump sql
+	sql, err := zw.Create("sqls.sql")
+	if err != nil {
+		return "", nil
+	}
+	_, err = sql.Write([]byte(e.ExecStmt.Text()))
+	if err != nil {
+		return "", err
+	}
+
+	// Dump bindings
+	recordSets, err = e.Ctx.(sqlexec.SQLExecutor).Execute(context.TODO(), "show bindings")
```

Comment:
Do we need to consider global bindings also?

Reply:
updated